qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/DenseVectorMappingUpdateIT.java

```diff
@@ -32,135 +36,181 @@
-                    if (useIndexSetting) {
+        String indexName = "test_index_type_change";
+        if (isOldCluster()) {
+            Request createIndex = new Request("PUT", "/" + indexName);
+            boolean useSyntheticSource = randomBoolean();
+
+            XContentBuilder payload = XContentBuilder.builder(XContentType.JSON.xContent()).startObject();
+            if (useSyntheticSource) {
+                payload.startObject("settings").field("index.mapping.source.mode", "synthetic").endObject();
+            }
+            payload.startObject("mappings");
+            payload.startObject("properties")
+                .startObject("embedding")
+                .field("type", "dense_vector")
+                .field("index", "true")
+                .field("dims", 8)
+                .field("similarity", "cosine")
+                .startObject("index_options")
+                .field("type", "hnsw")
+                .field("m", "16")
+                .field("ef_construction", "100")
+                .endObject()
+                .endObject()
+                .endObject()
+                .endObject()
+                .endObject();
+            createIndex.setJsonEntity(Strings.toString(payload));
+            client().performRequest(createIndex);
+            Request index = new Request("POST", "/" + indexName + "/_bulk/");
+            index.addParameter("refresh", "true");
+            index.setJsonEntity(generateBulkData(upgradedNodes, 8));
+            client().performRequest(index);
+        }
+
+        int expectedCount = 10;
+
+        assertCount(indexName, expectedCount);
+
+        if (isUpgradedCluster()) {
+            Request updateMapping = new Request("PUT", "/" + indexName + "/_mapping");
+            XContentBuilder mappings = XContentBuilder.builder(XContentType.JSON.xContent())
+                .startObject()
+                .startObject("properties")
+                .startObject("embedding")
+                .field("type", "dense_vector")
+                .field("index", "true")
+                .field("dims", 8)
+                .field("similarity", "cosine")
+                .startObject("index_options")
+                .field("type", "int8_hnsw")
+                .field("m", "16")
+                .field("ef_construction", "100")
+                .endObject()
+                .endObject()
+                .endObject()
+                .endObject();
+            updateMapping.setJsonEntity(Strings.toString(mappings));
+            assertOK(client().performRequest(updateMapping));
+            Request index = new Request("POST", "/" + indexName + "/_bulk/");
+            index.addParameter("refresh", "true");
+            index.setJsonEntity(generateBulkData(upgradedNodes, 8));
+            assertOK(client().performRequest(index));
+            expectedCount = 20;
+            assertCount(indexName, expectedCount);
+        }
+    }
+
+    private record Index(String type, Set<String> elementTypes) {}
```

Comment:
I am not sure how this will effectively allow skipping based on min node version.

Would this be better ? `Index(@Nullable String type, String elementType, boolean directIO, String minNodeVersionSupported)`

Unless you already have a nice way in mind to skip based on the combinations of element & index options.

Reply:
See the `getDimensions` method - which allows for feature checking based on the element type & type (and other aspects)